internal/client/oidc/client.go

```diff
@@ -60,105 +61,199 @@ var errRefreshAccessToken = fmt.Errorf("Failed refreshing access token")
+	return func(c *OIDCClient) {
+		c.authenticateCallback = callback
+	}
 }
 
 // NewClient constructs a new OidcClient, ensuring the token field is non-nil to prevent panics during authentication.
-func NewClient(httpClient *http.Client, tokensFile string) *OidcClient {
-	client := OidcClient{
-		tokens:        loadTokensFromFile(tokensFile),
-		tokensFile:    tokensFile,
-		httpClient:    httpClient,
-		oidcTransport: &oidcTransport{},
+func NewClient(httpClient *http.Client, oidcContextFile string, opts ...ClientOption) *OIDCClient {
+	client := &OIDCClient{
+		oidcContext:     loadOIDCContextFromFile(oidcContextFile),
+		oidcContextFile: oidcContextFile,
+		httpClient:      httpClient,
+		oidcTransport:   &oidcTransport{},
+
+		authenticateOpenBrowser: true,
 	}
 
-	// Ensure client.tokens is never nil otherwise authenticate() will panic.
-	if client.tokens == nil {
-		client.tokens = &oidc.Tokens[*oidc.IDTokenClaims]{}
+	for _, opt := range opts {
+		opt(client)
 	}
 
-	return &client
+	return client
 }
 
-func loadTokensFromFile(tokensFile string) *oidc.Tokens[*oidc.IDTokenClaims] {
-	if tokensFile == "" {
-		return nil
+func loadOIDCContextFromFile(oidcContextFile string) OIDCContext {
+	ret := OIDCContext{
+		Tokens: oidc.Tokens[*oidc.IDTokenClaims]{
+			Token: &oauth2.Token{},
+		},
 	}
 
-	ret := new(oidc.Tokens[*oidc.IDTokenClaims])
+	if oidcContextFile == "" {
+		return ret
+	}
 
-	contents, err := os.ReadFile(tokensFile)
+	contents, err := os.ReadFile(oidcContextFile)
 	if err != nil {
-		return nil
+		return ret
 	}
 
 	err = json.Unmarshal(contents, &ret)
 	if err != nil {
-		return nil
+		return ret
 	}
 
 	return ret
 }
 
-func saveTokensToFile(tokensFile string, tokens *oidc.Tokens[*oidc.IDTokenClaims]) error {
-	if tokensFile == "" {
+func saveOIDCContextToFile(oidcContextFile string, oidcContext OIDCContext) error {
+	if oidcContextFile == "" {
 		return nil
 	}
 
-	contents, err := json.Marshal(tokens)
+	contents, err := json.Marshal(oidcContext)
 	if err != nil {
 		return err
 	}
 
-	return os.WriteFile(tokensFile, contents, 0o644)
+	return os.WriteFile(oidcContextFile, contents, 0o600)
 }
 
 // GetAccessToken returns the Access Token from the OidcClient's tokens, or an empty string if no tokens are present.
-func (o *OidcClient) GetAccessToken() string {
-	if o.tokens == nil || o.tokens.Token == nil {
+func (o *OIDCClient) GetAccessToken() string {
+	o.oidcContextMu.Lock()
+	defer o.oidcContextMu.Unlock()
+
+	if o.oidcContext.Tokens.Token == nil {
 		return ""
 	}
 
-	return o.tokens.AccessToken
+	return o.oidcContext.Tokens.AccessToken
 }
 
 // GetOIDCTokens returns the current OIDC tokens, if any.
-func (o *OidcClient) GetOIDCTokens() *oidc.Tokens[*oidc.IDTokenClaims] {
-	return o.tokens
+func (o *OIDCClient) GetOIDCTokens() oidc.Tokens[*oidc.IDTokenClaims] {
+	o.oidcContextMu.Lock()
+	defer o.oidcContextMu.Unlock()
+
+	return o.oidcContext.Tokens
 }
 
+const earlyRefreshLeeway = 5 * time.Second
```

Comment:
@stgraber What do you think is a reasonable leeway for the early refresh of the access token? I think, this could be even bigger than the current 5 seconds, e.g. 30 seconds.